File: backend/app/api/schemas/product.py
"""
Pydantic schemas for Product data models.
Defines request/response schemas aligned with Odoo structure.
"""

from pydantic import BaseModel, Field, field_validator
from typing import Optional, List, Dict, Any, Annotated
from datetime import datetime
from bson import ObjectId
from pydantic_core import core_schema


class PyObjectId(str):
    """Custom type for MongoDB ObjectId compatible with Pydantic v2."""

    @classmethod
    def __get_pydantic_core_schema__(cls, source_type, handler):
        return core_schema.no_info_after_validator_function(
            cls.validate,
            core_schema.str_schema(),
        )

    @classmethod
    def validate(cls, v):
        if isinstance(v, ObjectId):
            return str(v)
        if isinstance(v, str):
            if not ObjectId.is_valid(v):
                raise ValueError("Invalid ObjectId")
            return v
        raise ValueError("Invalid ObjectId")


class ProductImage(BaseModel):
    """Product image with size variants."""
    image_id: str
    is_main: bool = False
    original_filename: str
    paths: Dict[str, str] = Field(default_factory=dict)  # size_256, size_512, etc.
    extracted_from: Optional[Dict[str, Any]] = None


class ProductDocument(BaseModel):
    """Product document (fiche technique, etc.)."""
    name: str
    path: str


class ProductSource(BaseModel):
    """Extraction source metadata."""
    source_id: str
    origin_file: str
    origin_file_type: str  # "pdf", "docx", "image", "video"
    page_number: Optional[int] = None
    extraction_type: str  # "text", "ocr", "vision", "audio"
    extracted_text: Optional[str] = None
    confidence_score: float = 0.0
    fields_extracted: List[str] = Field(default_factory=list)
    timestamp: datetime = Field(default_factory=datetime.utcnow)


class ManualEdit(BaseModel):
    """Manual edit history for a field."""
    field: str
    old_value: Any
    new_value: Any
    edited_date: datetime = Field(default_factory=datetime.utcnow)
    edited_by: Optional[str] = None


class ExtractionError(BaseModel):
    """Extraction error details."""
    error_type: str
    error_message: str
    field: Optional[str] = None
    timestamp: datetime = Field(default_factory=datetime.utcnow)


class ExtractionMetadata(BaseModel):
    """Metadata about product extraction."""
    extraction_date: datetime = Field(default_factory=datetime.utcnow)
    extraction_job_id: Optional[str] = None
    status: str = "raw"  # raw, enriched, validated, exported
    validation_date: Optional[datetime] = None
    validated_by: Optional[str] = None
    field_confidence_scores: Dict[str, float] = Field(default_factory=dict)
    manual_edits: List[ManualEdit] = Field(default_factory=list)
    errors: List[ExtractionError] = Field(default_factory=list)


class ProductBase(BaseModel):
    """Base product fields (for create/update)."""

    # Identifiers
    default_code: Optional[str] = None
    barcode: Optional[str] = None
    Code_EAN: Optional[str] = None  # Nom exact Odoo

    # Product Information
    name: Optional[str] = None
    type: str = "product"  # product, service, consumable
    active: bool = True
    is_published: bool = False

    # Classification
    categ_id: Optional[str] = None
    country_of_origin: Optional[str] = None

    # Manufacturer
    constructeur: Optional[str] = None
    refConstructeur: Optional[str] = None  # Nom exact Odoo

    # Descriptions
    description_courte: Optional[str] = None
    description_ecommerce: Optional[str] = None
    features_description: Optional[str] = None

    # Dimensions (mm and kg)
    length: Optional[float] = None
    width: Optional[float] = None
    height: Optional[float] = None
    weight: Optional[float] = None

    # Logistics
    hs_code: Optional[str] = None
    contient_du_lithium: bool = False

    # Pricing
    lst_price: Optional[float] = None
    taxes_id: List[str] = Field(default_factory=lambda: ["TVA 20%"])

    # Media
    images: List[ProductImage] = Field(default_factory=list)
    image_512: Optional[str] = None
    image_256: Optional[str] = None
    image_1920: Optional[str] = None
    image_1024: Optional[str] = None
    product_template_image_ids: List[int] = Field(default_factory=list)

    # Documents
    fiche_constructeur: Optional[ProductDocument] = None
    fiche_technique: Optional[ProductDocument] = None
    fiche_constructeur_nom: Optional[str] = None
    fiche_technique_nom: Optional[str] = None


class ProductCreate(ProductBase):
    """Schema for creating a new product."""
    pass


class ProductUpdate(BaseModel):
    """Schema for updating a product (all fields optional)."""

    # Identifiers
    default_code: Optional[str] = None
    barcode: Optional[str] = None
    Code_EAN: Optional[str] = None  # Nom exact Odoo

    # Product Information
    name: Optional[str] = None
    type: Optional[str] = None
    active: Optional[bool] = None
    is_published: Optional[bool] = None

    # Classification
    categ_id: Optional[str] = None
    country_of_origin: Optional[str] = None

    # Manufacturer
    constructeur: Optional[str] = None
    refConstructeur: Optional[str] = None  # Nom exact Odoo

    # Descriptions
    description_courte: Optional[str] = None
    description_ecommerce: Optional[str] = None
    features_description: Optional[str] = None

    # Dimensions
    length: Optional[float] = None
    width: Optional[float] = None
    height: Optional[float] = None
    weight: Optional[float] = None

    # Logistics
    hs_code: Optional[str] = None
    contient_du_lithium: Optional[bool] = None

    # Pricing
    lst_price: Optional[float] = None
    taxes_id: Optional[List[str]] = None

    # Documents
    fiche_constructeur_nom: Optional[str] = None
    fiche_technique_nom: Optional[str] = None

    # Metadata for tracking edits
    edited_by: Optional[str] = None


class Product(ProductBase):
    """Complete product schema (response from DB)."""

    id: PyObjectId = Field(default_factory=PyObjectId, alias="_id")

    # Extraction Metadata
    sources: List[ProductSource] = Field(default_factory=list)
    extraction_metadata: ExtractionMetadata = Field(default_factory=ExtractionMetadata)

    # Odoo Integration
    product_tmpl_id: Optional[int] = None
    odoo_product_tmpl_id: Optional[int] = None
    odoo_id: Optional[int] = None

    # Deduplication
    duplicate_group_id: Optional[str] = None
    is_master_record: bool = False
    merged_from: List[str] = Field(default_factory=list)

    # Timestamps
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    write_date: Optional[datetime] = None

    class Config:
        json_encoders = {ObjectId: str}
        populate_by_name = True


class ProductListResponse(BaseModel):
    """Response schema for product list."""
    products: List[Product]
    total: int
    page: int
    limit: int
    pages: int


class ProductResponse(BaseModel):
    """Response schema for single product."""
    product: Product
    message: Optional[str] = None

File: frontend/src/types/product.ts
/**
 * TypeScript types for Product data models.
 * Corresponds to backend Pydantic schemas.
 */

export interface OdooMatchInfo {
  odoo_id: number;
  score?: number;
  match_label?: string;
  /** True when applied automatically (score ≥ 90%). */
  auto: boolean;
  applied_fields: string[];
  matched_at?: string;
}

export interface ProductImage {
  image_id: string;
  is_main: boolean;
  original_filename: string;
  paths: {
    size_256?: string;
    size_512?: string;
    size_1024?: string;
    size_1920?: string;
  };
  extracted_from?: {
    file_path: string;
    page_number?: number;
    confidence: number;
  };
}

export interface ProductDocument {
  name: string;
  path: string;
}

export interface ProductSource {
  source_id: string;
  origin_file: string;
  origin_file_type: string;
  source_type?: string;        // "pdf" | "web_scrape" — set on new documents
  page_number?: number;
  extraction_type: string;
  extracted_text?: string;
  confidence_score: number;
  fields_extracted: string[];
  timestamp: string;
}

export interface ManualEdit {
  field: string;
  old_value: any;
  new_value: any;
  edited_date: string;
  edited_by?: string;
}

export interface ExtractionError {
  error_type: string;
  error_message: string;
  field?: string;
  timestamp: string;
}

export interface ExtractionMetadata {
  extraction_date: string;
  extraction_job_id?: string;
  status: 'raw' | 'enriched' | 'validated' | 'exported';
  validation_date?: string;
  validated_by?: string;
  field_confidence_scores: Record<string, number>;
  manual_edits: ManualEdit[];
  errors: ExtractionError[];
}

export interface Product {
  _id: string;

  // Identifiers
  default_code?: string;
  barcode?: string;
  Code_EAN?: string;  // Nom exact Odoo

  // Product Information
  name?: string;
  type: string;
  active: boolean;
  is_published: boolean;

  // Classification
  categ_id?: string;
  country_of_origin?: string;

  // Manufacturer
  constructeur?: string;
  refConstructeur?: string;  // Nom exact Odoo

  // Descriptions
  description_courte?: string;
  description_ecommerce?: string;
  features_description?: string;

  // Dimensions
  length?: number;
  width?: number;
  height?: number;
  weight?: number;

  // Logistics
  hs_code?: string;
  contient_du_lithium: boolean;

  // Pricing
  lst_price?: number;
  taxes_id: string[];

  // Media
  images: ProductImage[];
  image_512?: string;
  image_256?: string;
  image_1920?: string;
  image_1024?: string;
  product_template_image_ids: number[];
  /** Web-scraped image URLs: index 0 = main, rest = gallery */
  image_urls?: string[];
  source_url?: string;
  /** Web-scrape origin URLs (index 0 = primary). */
  scrape_source_urls?: string[];

  // Documents
  fiche_constructeur?: ProductDocument;
  fiche_technique?: ProductDocument;
  fiche_constructeur_nom?: string;
  fiche_technique_nom?: string;

  // Extraction Metadata
  sources: ProductSource[];
  extraction_metadata: ExtractionMetadata;

  // Odoo Integration
  product_tmpl_id?: number;
  odoo_product_tmpl_id?: number;
  odoo_id?: number;
  /** Link to the Odoo product this catalog product was matched/applied to. */
  odoo_match?: OdooMatchInfo;

  // Deduplication
  duplicate_group_id?: string;
  is_master_record: boolean;
  merged_from: string[];

  // Timestamps
  created_at: string;
  updated_at: string;
  write_date?: string;
}

export interface ProductListResponse {
  products: Product[];
  total: number;
  page: number;
  limit: number;
  pages: number;
}

/** One product inside a duplicate group (GET /products/duplicates/by-code). */
export interface DuplicateProduct {
  _id: string;
  name?: string;
  constructeur?: string;
  source_type?: string;
  status?: string;
  image_count: number;
}

export interface DuplicateGroup {
  default_code: string;
  count: number;
  products: DuplicateProduct[];
}

export interface DuplicatesByCodeResponse {
  groups: DuplicateGroup[];
  total_groups: number;
  page: number;
  limit: number;
  pages: number;
}

export interface ProductResponse {
  product: Product;
  message?: string;
}

export interface ProductUpdate {
  default_code?: string;
  barcode?: string;
  Code_EAN?: string;  // Nom exact Odoo
  name?: string;
  type?: string;
  active?: boolean;
  is_published?: boolean;
  categ_id?: string;
  country_of_origin?: string;
  constructeur?: string;
  refConstructeur?: string;  // Nom exact Odoo
  description_courte?: string;
  description_ecommerce?: string;
  features_description?: string;
  length?: number;
  width?: number;
  height?: number;
  weight?: number;
  hs_code?: string;
  contient_du_lithium?: boolean;
  lst_price?: number;
  taxes_id?: string[];
  fiche_constructeur_nom?: string;
  fiche_technique_nom?: string;
  edited_by?: string;
}

export interface ExtractionResult {
  message: string;
  filename: string;
  from_cache?: boolean;
  products_extracted: number;
  products: Array<{
    id: string;
    name?: string;
    default_code?: string;
  }>;
}

export interface ExtractionFileStatus {
  filename: string;
  /** pending → processing → done | cached | skipped | failed */
  status: 'pending' | 'processing' | 'done' | 'cached' | 'skipped' | 'failed';
  products: Array<{ id: string; name?: string; default_code?: string }>;
  product_count: number;
  from_cache: boolean;
  error: string | null;
}

export interface ExtractionJob {
  job_id: string;
  source: string;
  /** pending → running → done | failed */
  status: string;
  phase: string;
  phase_detail: string;
  created_at?: string;
  started_at?: string | null;
  finished_at?: string | null;
  total_files: number;
  processed_files: number;
  cached_files_count: number;
  failed_files_count: number;
  total_products: number;
  file_statuses: ExtractionFileStatus[];
  summary: {
    total_files: number;
    processed_successfully: number;
    cached: number;
    failed: number;
    total_products_extracted: number;
    images_processed?: number;
    images_associated?: number;
  } | null;
  error: string | null;
}
